### PR TITLE
25 create checkpoint app

### PR DESCRIPTION
**Pull request by KarnaushenkoVO**
## TRELLO

* [Main Trello ticket](https://trello.com/b/BJDdC443/mytrip)

## Code reviewers

- [x] @Ashiato
- [ ] @black-art
- [x] @hrytskivr
- [ ] @KarnaushenkoVO
- [x] @PanLeon
- [x] @yuriiZherebukh

### Second Level Review

- [ ] @byshovets
- [ ] @lhalam

## Summary of issue

Foreign key has been added to model.

## CHECK LIST
- [ ]  PR is reviewed manually again (to make sure you have 100% ready code)
- [ ]  All reviewers agreed to merge the PR
- [ ]  I've checked new feature as logged in and logged out user if needed
- [ ]  PR meets all conventions

### CODE REVIEW

File: myTrip/tests/checkpoint/test.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    """ Test for CRUD operation in checkpoint's view """
+
+    @classmethod
+    def setUpTestData(cls):
+        """Preconfig for test.
+        Include instance of Client to class as attribute
+        Create a model of trip
+        """
+        cls.client = Client()
+        new_trip = Trip(user_id = 3, title = "my_title", description = "some_cool_trip", status = 0)
```
Comment (hrytskivr) [inline myTrip/tests/checkpoint/test.py]: Consider testing with random data

File: myTrip/tests/checkpoint/test.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+                                    latitude=13,
+                                    title="new_ch",
+                                    description="coolsa_checkpoint",
+                                    source_url="my_sadurl",
+                                    position_number=2,
+                                    trip=new_trip)
+        new_checkpoint.save()
+
+    def test_get_status_success(self):
+        """Test for get operation with passed id ("""
```
Comment (hrytskivr) [inline myTrip/tests/checkpoint/test.py]: delete ' ( '
:D

File: myTrip/tests/checkpoint/test.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+        self.assertEqual(response.status_code, 200)
+
+    def test_get_status_not_found(self):
+        """Test for get operation with passed id (wrong)"""
+
+        response = self.client.get('/api/v1/trip/3/checkpoint/1000/')
+        self.assertEqual(response.status_code, 404)
+
+    def test_get_response_length(self):
+        """Test length of response object after get operation with passed id,  """
```
Comment (hrytskivr) [inline myTrip/tests/checkpoint/test.py]: delete ' , '

File: myTrip/tests/checkpoint/test.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+from trip.models import Trip
+from django.core.exceptions import ObjectDoesNotExist
+from checkpoint.models import Checkpoint
+from django.test import TestCase, Client
+
+
+JSON_LENGTH = 8
+
+class ViewTest(TestCase):
+    """ Test for CRUD operation in checkpoint's view """
```
Comment (hrytskivr) [inline myTrip/tests/checkpoint/test.py]: place a dot " . " in the end of docstring sentence

**Review by hrytskivr — COMMENTED** [review]
Looks good too mee

File: myTrip/checkpoint/migrations/0001_initial.py [hunk 5]
```diff
@@ -26,4 +26,4 @@ class Migration(migrations.Migration):
                 ('trip_id', models.IntegerField()),
             ],
         ),
-    ]
+    ]
```
Comment (lhalam) [inline myTrip/checkpoint/migrations/0001_initial.py]: add endline

File: myTrip/checkpoint/models.py [hunk 6]
```diff
@@ -15,16 +17,18 @@ class Checkpoint(models.Model):
     :argument description: text - description of checkpoint
     :argument position_number: int - ordinal number of checkpoint
     :argument source_url: url - url of the checkpoint's source
-    :argument trip: int - ToDo foreign key to Trip model
+    :argument trip: int - foreign key to Trip model
```
Comment (lhalam) [inline myTrip/checkpoint/models.py]: int to  Trip

File: myTrip/checkpoint/models.py [hunk 7]
```diff
@@ -15,16 +17,18 @@ class Checkpoint(models.Model):
...
     """
 
     longitude = models.FloatField()
     latitude = models.FloatField()
     title = models.CharField(max_length=20)
```
Comment (lhalam) [inline myTrip/checkpoint/models.py]: max to 70

File: myTrip/checkpoint/models.py [hunk 8]
```diff
@@ -15,16 +17,18 @@ class Checkpoint(models.Model):
...
     description = models.TextField()
     position_number = models.IntegerField()
-    source_url = models.URLField()
-    trip = models.IntegerField()
+    source_url = models.URLField(blank=True)
+    trip = models.ForeignKey(Trip, on_delete=models.CASCADE)
+    created = models.DateTimeField()
```
Comment (lhalam) [inline myTrip/checkpoint/models.py]: create_at

File: myTrip/checkpoint/models.py [hunk 9]
```diff
@@ -15,16 +17,18 @@ class Checkpoint(models.Model):
...
+    last_modified = models.DateTimeField()
```
Comment (lhalam) [inline myTrip/checkpoint/models.py]: update_at

File: myTrip/checkpoint/models.py [hunk 10]
```diff
@@ -38,10 +42,17 @@ def to_dict(self):
...
         """
-
+        return {"id": self.id,
+                "longitude": self.longitude,
+                "latitude": self.latitude,
+                "title": self.title,
+                "description": self.description,
+                "source_url": self.source_url,
+                "position_number": self.position_number,
+                "trip": self.trip.id}
```
Comment (lhalam) [inline myTrip/checkpoint/models.py]: trip => trip_id

File: myTrip/checkpoint/models.py [hunk 11]
```diff
@@ -57,38 +68,44 @@ def create(longitude, latitude, title, description, source_url, position_number,
             source_url (url):  url of the checkpoint's source.
             trip (int):  ToDo foreign key to Trip model.
         Returns:
-            new checkpoint object
+            Object<Checkpoint>: Object of Checkpoint.
         """
 
+        new_checkpoint = Checkpoint()
```
Comment (lhalam) [inline myTrip/checkpoint/models.py]: new_checkpoint => checkpoint

File: myTrip/checkpoint/models.py [hunk 12]
```diff
@@ -57,38 +68,44 @@ def create(longitude, latitude, title, description, source_url, position_number,
...
-            new_object.save()
+            new_checkpoint.save()
+            return new_checkpoint
         except FieldError:
             return None
-        return new_object
 
     @staticmethod
-    def get_by_id(checkpoint_id):
+    def get_by_id(checkpoint_id, trip_id):
```
Comment (lhalam) [inline myTrip/checkpoint/models.py]: split into 2 methods

File: myTrip/checkpoint/models.py [hunk 13]
```diff
@@ -125,16 +142,17 @@ def update(self,
...
         if source_url:
             self.source_url = source_url
-        if longitude:
-            self.trip = trip
+        if trip:
+            self.trip = Trip.get_by_id(trip)
+        self.last_modified = datetime.now()
         try:
             self.save()
             return self
```
Comment (lhalam) [inline myTrip/checkpoint/models.py]: delete returns

File: myTrip/checkpoint/models.py [hunk 14]
```diff
@@ -125,16 +142,17 @@ def update(self,
...
-        except ProtectedError:
+        except models.ProtectedError:
             return None
```
Comment (lhalam) [inline myTrip/checkpoint/models.py]: pass

File: myTrip/checkpoint/models.py [hunk 15]
```diff
@@ -125,16 +142,17 @@ def update(self,
...
 
     @staticmethod
-    def delete(checkpoint_id):
+    def delete(checkpoint_id,trip_id):
```
Comment (lhalam) [inline myTrip/checkpoint/models.py]: only checkpoint_id

File: myTrip/checkpoint/models.py [hunk 16]
```diff
@@ -143,9 +161,9 @@ def delete(checkpoint_id):
             True if deleting was successful
             False if deleting wasn't complete
         """
-        checkpoint = Checkpoint.objects.get(id = checkpoint_id)
+        checkpoint = Checkpoint.get_by_id(checkpoint_id,trip_id)
```
Comment (lhalam) [inline myTrip/checkpoint/models.py]: Checkpoint.objects.get

**Review by lhalam — COMMENTED** [review]
ToDo

File: myTrip/tests/checkpoint/unit_test/test.py [hunk 17]
```diff
@@ -0,0 +1,178 @@
+"""Testing module for checkpoint views."""
+
+import json
+from datetime import datetime
```
Comment (hrytskivr) [inline myTrip/tests/checkpoint/unit_test/test.py:4]: json and datetime are standart python libs, so after them should be space, before djangos imports

File: myTrip/checkpoint/models.py [hunk 18]
```diff
@@ -17,18 +17,18 @@ class Checkpoint(models.Model):
...
     latitude = models.FloatField()
-    title = models.CharField(max_length=20)
+    title = models.CharField(max_length=70)
     description = models.TextField()
     position_number = models.IntegerField()
     source_url = models.URLField(blank=True)
     trip = models.ForeignKey(Trip, on_delete=models.CASCADE)
-    created = models.DateTimeField()
-    last_modified = models.DateTimeField()
+    create_at = models.DateTimeField(editable=False)
```
Comment (hrytskivr) [inline myTrip/checkpoint/models.py]: Suggestion:
created_at
modified_at

why use editable false?

**Review by hrytskivr — COMMENTED** [review]
Check this out

**Review by hrytskivr — APPROVED** [review]
Good :)

File: myTrip/tests/checkpoint/unit_test/test.py (hunk 17 above)
Comment (KarnaushenkoVO) [reply myTrip/tests/checkpoint/unit_test/test.py:4]: no, only after my own modules

**Review by yuriiZherebukh — APPROVED** [review]

**Review by oshtohryn — APPROVED** [review]

**Pull request merged**